File: leetcode/golang/3_longest_substring_without_repeating_characters.go
func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func lengthOfLongestSubstring(s string) int {
    sMap := make(map[byte]int)
    res := 0
    for i, j := 0, -1; i < len(s); i++ {
        if _, ok := sMap[s[i]]; ok {
            j = max(j, sMap[s[i]])
        }
        sMap[s[i]] = i;
        res = max(res, i - j)
    }
    return res
}

File: leetcode/c++/3_longest_substring_without_repeating_characters.cpp
// https://leetcode.cn/problems/longest-substring-without-repeating-characters

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> char_map;
        int res = 0;
        for (int i = 0, j = -1; i < s.length(); i++) {
            if (char_map.find(s[i]) != char_map.end()) {
                j = max(j, char_map[s[i]]);
            }
            char_map[s[i]] = i;
            res = max(res, i - j);
        }
        return res;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(1)